File: code/437.path-sum-iii.php
<?php
/*
 * @lc app=leetcode id=437 lang=php
 *
 * [437] Path Sum III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {
    private int $count = 0;
    private int $targetSum;

    /**
     * @param TreeNode $root
     * @param Integer $targetSum
     * @return Integer
     */
    function pathSum($root, $targetSum) {
        if ($root == null) {
            return 0;
        }

        $this->targetSum = $targetSum;
        $this->sum($root, 0, 0);

        return $this->count;
    }

    function sum($root, $current, $deps) {
        $sum = $current + $root->val;
        if ($sum === $this->targetSum) {
            $this->count++;
        }

        if ($root->left != null) {
            $this->sum($root->left, $sum, $deps + 1);
            if ($deps < 1) {
                $this->sum($root->left, 0, 0);
            }
        }
        if ($root->right != null) {
            $this->sum($root->right, $sum, $deps + 1);
            if ($deps < 1) {
                $this->sum($root->right, 0, 0);
            }
        }
    }
}
// @lc code=end

File: code/437.path-sum-iii.go
package code

/*
 * @lc app=leetcode id=437 lang=golang
 *
 * [437] Path Sum III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	return pathSumRec(root, targetSum) + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}

func pathSumRec(node *TreeNode, targetSum int) int {
	if node == nil {
		return 0
	}

	count := 0
	remain := targetSum - node.Val
	if remain == 0 {
		count++
	}

	count += pathSumRec(node.Left, remain)
	count += pathSumRec(node.Right, remain)

	return count
}

// @lc code=end
